oeps/oep-0045-arch-ops-and-config.rst

```diff
@@ -0,0 +1,175 @@
+
+.. code-block:: text
+
+  settings
+  ├── __init__.py
+  ├── required.py
+  └── defaults.py
+
+
+*  ``__init__.py`` - Sourcing our config from this file within the settings directory takes advantage of django defaults and means that settings will be picked up automatically without needing to specify ``--settings`` anywhere. This entry point would import ``required.py``, ``defaults.py``, and the code to override both from a config file.
+*  ``required.py`` - all settings which are required to run and do not have a reasonable production-ready default, e.g. LMS_BASE_URL which will be different per environment.
+*  ``defaults.py`` - other settings which will have production-ready defaults
+
+The settings defined in ``required.py`` and ``defaults.py`` files are mutually exclusive, representing all application specific settings as well as installed library settings whose values either must be provided or whose defaults are not considered production-ready.
+
+``required.py`` variables must be overridden by operators.  The application will check that operators provided these values, and will not start unless they are set. This allows operators to fail fast rather than finding out about an unset value when users exercise those breaking codepaths. Application developers are encouraged to keep the list of required settings to a minimum.
+
+This new settings structure obviates the need for any other python files in the settings directory (such as ``devstack.py``, ``test.py``, etc). The values currently set in those files should be moved to a corresponding ``devstack.yml``, ``test.yml``, etc in the same settings directory. This gives developers and operators more consistency across environments since the same code paths are being executed with different values.
+
+
+**Config file**
+
+Applications will be configured by a yaml file containing all of the settings variable overrides specified by the operator (including both required settings and secrets as well as default value overrides). The file is made known to the application by an environment variable, ``<APPNAME>_CFG_PATH``, with the path to the file. Versions of this config yaml may be provided in the application repo for certain environments such as development and test. However, for all other environments (e.g. production), the file will need to be managed elsewhere.
+
+Since defaults are provided by the application, many smaller deployments should not need to do much more than provide the required settings to operate. For development environments the config will likely change the defaults to more development appropriate values, e.g. debug settings, log levels, email settings, etc.
+ 
+**Config file generation & management**
+
+Due to the varied needs and processes of different operators, how the config files are created, managed, or otherwise end up on the server is up to the operator and will depend greatly on their deployment strategy.
```

Comment:
Will we be generating test.yml?

Reply:
While the deployment processes will often vary, does (or _should_) the creation vary?

If we're forcing certain values to be set by operators (and enforcing via early failure), it seems like we ought to provide a canonical way to create the files, or at least provide a blessed reference implementation.

The actual implementation would vary, but I'm imaging at least a helper of sorts, something in the spirit of:
- https://github.com/edx/configuration/blob/eec2b4ae2fdf88fd17aabfc11a308951b243f9e4/util/install/generate-passwords.sh

Reply:
@nedbat I think that applications should define a `test.yml` that can be used, I would not expect to need to generate them.

@stvstnfrd I don't have strong opinion here.  My thought was that it is so simple that no reference or helper would be required.  Value defined in `required.py' or `default.py` maps 1-to-1 to the value you stick in your yml file to override it.  Interested to hear more about what the helper would actually do

Reply:
@bderusha I guess I don't know either :) I'll probably have a better idea upon seeing potential implementations.
And no, I'm not expecting that at this point, just _slightly_ pushing back on the idea of deferring the responsibility permanently :)

My ears perk up whenever I hear, "operators can figure this out themselves", as those were often the areas that bit us most frequently :P

It's probably less of an issue for us than it will be for the community.
- I assume we'll generally make the change to `{required,default}.py` immediately in tandem with the change to the `${env}.yml` (whether that's in a private repo or what not).
- On the other hand, community operators are going to "inherit" the changes to `{required,default}.py` weeks/months/years later, and then will need to reconcile the two files, line by line (?? since it won't be a simple `diff`, since we're comparing `py` and `yml` files) or while auditing the `CHANGELOG`.
  - Of course, it won't be the "worst thing ever" :tm: , but it seems like an extra tedium to incur upon operators, that we won't feel ourselves.

Off the top of my head (again, I'm sure seeing sample implementations will help my mental model), I could imagine a script that auto-merges/reconciles the `py` files with the `yml`. Something like:

```sh
./upgrade-config --input ironwood.yml --output juniper.yml
# defaults.py and required.py are implied default inputs
```

So assuming input files like:

```py
## required.py
ADMIN_EMAIL = YAML_CONFIG['ADMIN_EMAIL']
NEW_TOGGLE_ADDED_IN_JUNIPER = YAML_CONFIG['NEW_TOGGLE_ADDED_IN_JUNIPER']
```

```py
## defaults.py
LMS_BASE = YAML_CONFIG['LMS_BASE']
ANOTHER_TOGGLE_ADDED_IN_JUNIPER = YAML_CONFIG['ANOTHER_TOGGLE_ADDED_IN_JUNIPER']
```

```yml
## ironwood.yml
ADMIN_EMAIL: 'admin@example.com'
LMS_BASE: 'localhost:18000'
OLD_TOGGLE_REMOVED_IN_JUNIPER: true
```

Might yield an output file like:

```yml
## juniper.yml
## This file was automatically generated by: `upgrade-config`

## Deprecated settings
## The following are now unsupported in this release:
# OLD_TOGGLE_REMOVED_IN_JUNIPER: true

## Required settings
## These must all be set:
ADMIN_EMAIL: 'admin@example.com'
# NEW_TOGGLE_ADDED_IN_JUNIPER: 'recommended default value'

## Default overrides
## These are optional overrides:
LMS_BASE: 'localhost:18000'
# ANOTHER_TOGGLE_ADDED_IN_JUNIPER: 'default to the value in `defaults.py`'
```

An operator would still need to tweak the output file (to update new required values), but I can imagine some process like this making it easier and more obvious to automatically see:
- which new required settings have been added
- which new optional overrides are available
- which settings have been deprecated and are no longer supported

I'm sure there would be deployments where this isn't as useful, but I'd expect it to remove busy work and mistakes for many operators.

Reply:
Agree with @stvstnfrd last point sounds very reasonable.